### DISCUSSION TITLE
/blog/2022/05/20/marginalia/

### DISCUSSION THREAD

**Discussion by andrewheiss** [andrewheiss/ath-quarto#48] [Blog comments]
Marginalia: A guide to figuring out what the heck marginal effects, marginal slopes, average marginal effects, marginal effects at the mean, and all these other marginal things are | Andrew Heiss 
 https://www.andrewheiss.com/blog/2022/05/20/marginalia/

**Comment by andrewheiss** [comment]
```header
Originally posted by Niklas Cypris on 2022-05-26 17:49:39
````
Super interesting! This blog post is helping me a lot with my
PhD-related research. Thank you a lot :)

**Comment by andrewheiss** [comment]
```header
Originally posted by Felix Thoemmes on 2022-06-10 02:11:44
````
Wonderful blog post, as usual. One may &quot;trick&quot; emmeans in also
spitting out AMEs by forming quadratic and interaction terms by hand.
Example here:\
[https://github.com/rvlenth/\...](https://github.com/rvlenth/emmeans/issues/45)

**Comment by andrewheiss** [comment]
```header
Originally posted by Przemek on 2022-07-16 12:38:54
````
Thanks for this post - extremely useful. However, I still don\'t know
which of these approaches is used by another quite popular package,
&quot;effects&quot; by John Fox. Any idea?

**Comment by andrewheiss** [comment]
```header
Originally posted by Tiny van Boekel on 2022-07-31 08:30:51
````
Even if you are not super smart (I think you are!), you are a super
teacher! As a non-statistician I could not get my head around
differences between marginal and conditional effects and contrasts and
all that but this post clarifies a lot in a very clever way, very
inspirational! One question in addition: did you ever take a look at the
hypothesis function of brms? In this blog post
[https://sometimesir.com/pos\...](https://sometimesir.com/posts/2020-02-06-how-to-calculate-contrasts-from-a-fitted-brms-model/ &quot;https://sometimesir.com/posts/2020-02-06-how-to-calculate-contrasts-from-a-fitted-brms-model/&quot;)
Matti Vuorre calls it an underappreciated method of the brms package. It
can be used to calculate contrasts as he shows. I wonder whether you
have any experience with this brms function?

**Comment by andrewheiss** [comment]
```header
Originally posted by Arthur Albuquerque on 2022-09-09 22:38:32
````
Andrew, this is a wonderful post. I\'d like to make a few comments about
the &quot;Where this subtle difference really matters&quot; section.

I\'m not sure if comparing the point estimates and 95% CIs of the AME
vs. MEM are informative. I believe these are different estimands (AME =
marginal \[population-averaged\], while MEM is conditional). Using
Daniel et al.\'s vocabulary, you are comparing apples to oranges:

&quot;Conditional and marginal odds ratios (likewise hazard ratios) are like
apples and oranges. It is true that, in the absence of confounding etc.,
the standard error of the maximum likelihood estimator (MLE) of a
conditional odds ratio is at least as large as the standard error of the
corresponding MLE of a marginal odds ratio (similarly for hazard ratios)
(Ford et al., 1995; Robinson &amp; Jewell, 1991). This statement is arguably
irrelevant, however, since it says that the standard error of an
estimator of an orange is at least as large as the standard error of an
estimator of an apple&quot;\
[https://onlinelibrary.wiley\...](https://onlinelibrary.wiley.com/doi/full/10.1002/bimj.201900297)

What do you think?

**Reply by andrewheiss** [reply]
```header
Originally posted on 2022-09-15 19:53:51
````
I would agree :). The AME and MEM are different estimands, which is why
they\'re so different in the results. In real life you probably
shouldn\'t calculate both and compare them to each other

**Reply by andrewheiss** [reply]
```header
Originally posted by Arthur Albuquerque on 2022-09-15 20:16:10
````
I\'d say you could calculate both, but should clearly state the
parameter of interest is completely different in each one.

**Comment by mvuorre** [comment]
>Public work is also valuable for another more...

This really resonated with me. Putting your work out there is always valuable, sometimes in surprisingly positive ways. (Given that it is not in error, that is 😆). Cheers and keep 'em coming.

**Reply by andrewheiss** [reply]
Even being wrong in public is useful! I'm a huge fan of @ASKurz's [philosophy here](https://twitter.com/SolomonKurz/status/1572218190159134720) that there's benefit to trying things out (and failing) in public

<img width="594" alt="image" src="https://user-images.githubusercontent.com/73663/204639594-2845fd94-c35f-4042-b4e4-5d8c064bf9bd.png">

**Reply by ASKurz** [reply]
Cheers!

**Reply by mvuorre** [reply]
Heck yeah. We all learn from each other.

**Comment by nahorp** [comment]
Thanks for a *super* helpful blogpost Andrew! Have used `emmeans` exclusively but discovered `marginaleffects` recently and this has helped me to understand the various differences. Now to go and figure out all this in a Bayesian context!

**Reply by andrewheiss** [reply]
This might be helpful for Bayesian models! https://www.andrewheiss.com/blog/2022/11/29/conditional-marginal-marginaleffects/

**Comment by juanpide** [comment]
Hello.

You say "a marginal effect is only a partial derivative".
But what if we have multiple covariates? What do we do we the other variables?

**Reply by andrewheiss** [reply]
You hold them constant in different ways. The post explains this, starting at this section here: https://www.andrewheiss.com/blog/2022/05/20/marginalia/#marginaleffectss-and-emmeanss-philosophies-of-averaging 

There's also a summary table at the end explaining how to handle covariates in different ways https://www.andrewheiss.com/blog/2022/05/20/marginalia/#tldr-overall-summary-of-all-these-marginal-effects-approaches

Also check out the slider/switch/mixing board analogy here: https://www.andrewheiss.com/blog/2022/05/20/marginalia/#regression-sliders-switches-and-mixing-boards

**Comment by rvlenth** [comment]
Just briefly commenting. I am the developer of **emmeans**, and a user referred me to this. I have just skimmed it so far. I find it interesting, and at least on the surface I think you are mostly correct in explaining distinctions between the two packages.

I am not surprised to find that, with an econ emphasis, you find **marginaleffects** preferable, and that is for good reason given that it more easily calculates the sorts of things needed in that application area. That's because, as stated in its vignettes, **emmeans** emphasizes interpretation of experimental data rather than observational data, which I believe is the primary focus of econometric analyses.

In the summary, there are several topics where you say such analyses are not supported by **emmeans**. I think that in at least some of those cases, what you really are saying is that you don't know how to do it with **emmeans**. For example, the `ref_grid()` function (and therefore `emmeans()` and `emtrends()`) has an optional `counterfactuals` argument that causes the reference grid to be constructed appropriately for counterfactuals analysis. The `emmeans()` function supports a `weights` argument for weighting the marginal means differently than the default equal weighting, and very often with observational data, one should use some kind of proportional weighting so as to reflect what is going on in the population. 

I'm not saying that (**emmeans** is just as good a choice for your purposes, nor even that I am sure that it supports all of your scenarios. But I do think some things that are "harder to do" are characterized as "impossible."

**Reply by andrewheiss** [reply]
Hello! And thanks for {emmeans}! It's a great package and I still use it in addition to {marginaleffects—I} have an experimental project I'm working on right now where we're finding estimated marginal means with {emmeans}!

I wasn't aware that `ref_grid()` had added counterfactual support. I wrote this in May 2022, and looking at the {emmeans} news file it looks like the counterfactual argument was added a few months after that. I'll update the post!

From my read on it (and [Vincent's vignette over at {marginaleffects}](https://marginaleffects.com/articles/alternative_software.html)), really the main difference is what you point out—{emmeans} is perfect for estimated marginal means and anything else where you plug in specific values to a model, like in experiments, while {marginaleffects} can do the econometrics-style generate-unit-level-predictions-and-then-marginalize approach ([like the contrasts section here](https://marginaleffects.com/articles/alternative_software.html#contrasts)).

**Comment by rvlenth** [comment]
Andrew, Thanks for the response and for pointing me to Vincent's interesting comments and comparisons. Those are thoroughly done. The one thing I see not emphasized there is **emmeans**'s default to equally-weighted marginal means of predictions. That is an important distinction and reflects my own emphasis on experimental studies. It really has a lot to do with the types of analyses where we try to characterize a population (observational study) versus what happens if we tweak some dials in the system (experimental study).

**Comment by tjmahr** [comment]
the light switch (categorical) versus dimmer (continuous) makes me so happy

**Comment by rbalshaw** [comment]
Andrew - very much appreciate both the open-process spirit and the wonderful examples. In your interpretations of the coefficients produced by the glm() with link = logit, I think you are correct to caution they must be interpreted on the logit scale.

However, I wonder if the coefficients in the outputs of the marginaleffects functions may be on the same scale?  That is, I think you still need to be cautious and exponentiate the coefficients to then interpret them as odds-ratios... and from there, as percentages.

It's a convenient and sometimes confusing quirk of the exponential function that when a coefficient from a logistic regression model is "close" to 0 (on the scale of the linear predictor), it can be read as a close approximation to the percentage increase in the odds of success (if coef is positive) or decrease (if coef is negative).  The slope of exp(x) is 1 at x=0, and is "close to linear" for a short range of x values around 0.

I struggle with explanations, so here's a few versions of what I'm trying to say.

(a) If your estimated beta has coefficients that are "close to 0" (say, between -0.15 and 0.15), exp(beta) will be "close to" the value of (1 - OR)*100%.

(b) A specific example: if the coef estimate (on the scale of the linear predictor) was -0.068 --> exp(-0.068) = 0.934.  So, we'd often interpret this as a "7% decrease in the odds of success",  Similarly, if the coef was 0.125, we could immediate interpret this as "approximately a 13% increase in the odds of success" as exp(0.125) = 13.3% (note the approximation is already getting a bit wonky as we move away from a coef near 0).

Two caveats:
(i) I've been wrong twice already today, and wouldn't be surprised if this were the third.  Maybe the emmeans and marginaleffects packages operate differently than I think.  I did a bit of checking but did not deep dive.  Too many interesting rabbit holes to go down!
(ii) Sorry I'm late to the discussion.

**Reply by andrewheiss** [reply]
Hi! It turns out that I wasn't getting notifications from GitHub about new comments, so I missed this!

{marginaleffects} returns values on the percentage point scale by default, which is why it's such a powerful and neat package for understanding and interpreting models. There's [a detailed logit vignette at the {marginaleffects} documentation](https://marginaleffects.com/vignettes/logit.html) about a bunch of different estimands you can get from logistic regression models.

Here's a quick example with the penguins data, with a binary indicator showing if a penguin is a Gentoo or not:

```r
library(tidyverse)
library(marginaleffects)
library(palmerpenguins)
library(parameters)

penguins <- penguins |> 
  drop_na(sex) |> 
  mutate(is_gentoo = species == "Gentoo")

model <- glm(
  is_gentoo ~ bill_length_mm + body_mass_g,
  data = penguins,
  family = binomial(link = "logit")
)
```

We can get the log odds values and odds ratios from the raw coefficients:

```r
parameters(model)
#> Parameter      | Log-Odds |       SE |           95% CI |     z |      p
#> ------------------------------------------------------------------------
#> (Intercept)    |   -32.40 |     4.71 | [-42.81, -24.21] | -6.88 | < .001
#> bill length mm |     0.09 |     0.06 | [ -0.03,   0.21] |  1.51 | 0.132 
#> body mass g    | 6.36e-03 | 8.57e-04 | [  0.00,   0.01] |  7.42 | < .001
#> 
#> The model has a log- or logit-link. Consider using `exponentiate =
#>   TRUE` to interpret coefficients as ratios.

parameters(model, exponentiate = TRUE)
#> Parameter      | Odds Ratio |       SE |       95% CI |     z |      p
#> ----------------------------------------------------------------------
#> (Intercept)    |   8.48e-15 | 3.99e-14 | [0.00, 0.00] | -6.88 | < .001
#> bill length mm |       1.09 |     0.06 | [0.97, 1.23] |  1.51 | 0.132 
#> body mass g    |       1.01 | 8.63e-04 | [1.00, 1.01] |  7.42 | < .001
```

A 1-mm increase in bill length is associated with a 0.09 increase in the log odds of being a Gentoo (whatever that means), or it is associated with a 9% increase in the likelihood of being a gentoo (`exp(0.09)` = 1.09 OR). Those two values (β and e<sup>β</sup>) are all on weird logit-related scales, and that's all fine and normal.

{marginaleffects} lets you find more interpretable percentage-point-scale values. For instance, if we hold body mass constant and vary bill length, we can see how the probability of being a Gentoo changes across all the values of bill length. The slope is shallow below 40 mmm and is steeper after 50 mm:

```r
plot_predictions(model, condition = "bill_length_mm")
```

![](https://i.imgur.com/hgAVoGb.png)

We can find the exact slope of that predicted line at different values:

```r
slopes(model, newdata = datagrid(bill_length_mm = c(35, 45, 55)), variables = "bill_length_mm")
#> 
#>            Term bill_length_mm Estimate Std. Error    z Pr(>|z|)   S    2.5 %  97.5 %
#>  bill_length_mm             35  0.00602    0.00188 3.21  0.00134 9.5  0.00234 0.00969
#>  bill_length_mm             45  0.01200    0.00836 1.44  0.15123 2.7 -0.00439 0.02838
#>  bill_length_mm             55  0.01928    0.01706 1.13  0.25847 2.0 -0.01416 0.05272
#> 
#> Columns: rowid, term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, bill_length_mm, predicted_lo, predicted_hi, predicted, body_mass_g, is_gentoo 
#> Type:  response
```

That estimate column shows percentage points. For a penguin with a bill length of 35 mm, a 1-mm increase is associated with a 0.6 percentage point (`0.00602 * 100`) increase in the probability of being a Gentoo. For a penguin with a bill length of 55 mm, a 1-mm increase is associated with a nearly 2 percentage point (`0.01928 * 100`) increase in the probability of being a Gentoo. Those three slopes are what you see at the vertical lines here:

```r
plot_predictions(model, condition = "bill_length_mm") +
  geom_vline(xintercept = c(35, 45, 55), linetype = "dotted") +
  scale_y_continuous(labels = scales::label_percent())
```

![](https://i.imgur.com/p9vnsUL.png)

**Comment by rbalshaw** [comment]
Andrew - thanks for getting back to me.  If I recall my state of mind when I made the comment/asked the questions, I was wanting to confirm my understanding of the terminology you used.

Working with clients who sometimes want risk differences from linear probability models, sometimes conditional, sometimes marginal risk differences from a logistic regression model, as well as teaching students the basics of logistic regression and having to explain the "handy but sometimes awkward" coincidence that an logistic regression coefficient of 0.09 means both a 9% increase in the log-odds of "success" and a 9% increase in the odds, as exp(0.09) = 1.09 (approx), has left me inclined to double check that I'm on the same page with them.

Thanks for clarifying.

In any case, I've since been reading some of the tutorials and experimenting a bit with other elements of the package and all I can say is "Amazing work!"   The breadth of the package is going to take some time to absorb.  A veritable zoo, indeed!

Quick question – with apologies, as I'm sure this is answered elsewhere – but what method is used to achieve SEs for avg_comparisons()?  Does the default method depend on the nature of the estimand?  For example, in the case of a simple linear model, is this just the SE of the contrast, calculated as we've always done it since the dawn of SAS LSMEANS?  And, what about in the case of the g-formula?  Bootstrapping?  Some form of sandwich estimator or delta method?

That question probably betrays my level of unfamiliarity with the package.  RTFM might be a completely reasonable response!

Again, thanks for your response – and your work on this package!

Cheers,

Rob

________________________________

**Reply by andrewheiss** [reply]
Ah, there's a whole page on standard error calculations :) https://marginaleffects.com/vignettes/uncertainty.html

By default all the marginaleffects functions use the delta method, but you can use the `vcov` argument to change them to to whatever you want (sandwich, robust Stata-like things, bootstrapping, etc)

**Comment by rbalshaw** [comment]
Ah! I should have known you'd have this covered!

In the non-inferiority testing context, people get their shirts in a twist about width of confidence intervals.  We're often a bit laissez faire when it comes to potential overestimation of variances when doing superiority testing – it's more conservative that way, after all! - but wider confidence intervals could be seen as anti-conservative when trying to show "pretty much the same".

I'm just wallowing in that literature for a non-inferiority trial with binary outcomes, and a non-inferiority marging expressed as a difference in proportions rather than something more polite like an OR.

Thanks again for the package and the assistance!

Rob
________________________________

**Comment by TranHung93** [comment]
Dear Andrew, 
Thanks for your great blog. I've learnt a lot from you. 

There's just a few typos I think. I still can understand that part but you might like to fix it if you have time. In the line "In Western Europe, the AME of corruption is `r m_logit_fancy_mfx$public_sector_corruption_2$est`" , I think it should be `r m_logit_fancy_mfx$public_sector_corruption_5$est` instead. Otherwise, it might show the value of Latin America. Simililarly, the middle east value need be changed to `r m_logit_fancy_mfx$public_sector_corruption_3$est` . 

Regards,
Hung

**Comment by sjpierce** [comment]
This is an extraordinarily useful post (including the additional discussion in the comments). I had been using **emmeans** (an exceptionally useful package) and was aware of **marginaleffects** but had not really delved into it much. I've learned quite a bit about both of those packages here. I will be more thoughtful about when to use each of them in the future. I suspect that the context of experimental vs observational studies should indeed play a role in that decision. 

I'm going to point out this blog post to my colleagues and encourage them to read it.

**Comment by jonong03** [comment]
Thank you for writing this, it helps a lot on connecting the dots!

**Comment by lricketts** [comment]
Thank you for putting together this wonderful post! It has been very instructive and will be a useful reference going forward. I will definitely be following your other work!
